twoliter/embedded/imghelper

```diff
@@ -1,6 +1,150 @@
 #!/usr/bin/env bash
-
 # shellcheck disable=SC2034 # variables are used through parameter substitution
+
+DATA_IMAGE_SUFFIX="-data"
+BOOT_IMAGE_SUFFIX="-boot.ext4.lz4"
+VERITY_IMAGE_SUFFIX="-root.verity.lz4"
+ROOT_IMAGE_SUFFIX="-root.ext4.lz4"
+
+set_image_names() {
+  local filename_prefix
+  local -n os_image_name data_image_name
+  local -n boot_image_name verity_image_name root_image_name
+  filename_prefix="${1:?}"
+  os_image_name="${2:?}"
+  data_image_name="${3:?}"
+  boot_image_name="${4:?}"
+  verity_image_name="${5:?}"
+  root_image_name="${6:?}"
+
+  os_image_name="${filename_prefix}"
+  data_image_name="${filename_prefix}${DATA_IMAGE_SUFFIX}"
+  boot_image_name="${filename_prefix}${BOOT_IMAGE_SUFFIX}"
+  verity_image_name="${filename_prefix}${VERITY_IMAGE_SUFFIX}"
+  root_image_name="${filename_prefix}${ROOT_IMAGE_SUFFIX}"
+}
+
+sanity_checks() {
+  local output_fmt partition_plan ovf_template uefi_secure_boot
+  output_fmt="${1:?}"
+  partition_plan="${2:?}"
+  ovf_template="${3:?}"
+  uefi_secure_boot="${4:?}"
+  case "${output_fmt}" in
+  raw | qcow2 | vmdk) ;;
+  *)
+    echo "unexpected image output format '${output_fmt}'" >&2
+    exit 1
+    ;;
+  esac
+
+  case "${partition_plan}" in
+  split | unified) ;;
+  *)
+    echo "unexpected partition plan '${partition_plan}'" >&2
+    exit 1
+    ;;
+  esac
+
+  # Fail fast if the OVF template doesn't exist, or doesn't match the layout.
+  if [ "${output_fmt}" == "vmdk" ]; then
+    if [ ! -s "${ovf_template}" ]; then
+      echo "required OVF template not found: ${ovf_template}" >&2
+      exit 1
+    fi
+
+    if [ "${partition_plan}" == "split" ]; then
+      if ! grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Missing data disk in OVF template, which is required for 'split' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${partition_plan}" == "unified" ]; then
+      if grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Incorrect data disk in OVF template, which is not supported for 'unified' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${uefi_secure_boot}" == "yes" ]; then
+      if ! grep -Fq '{{DB_CERT_DER_HEX}}' "${ovf_template}"; then
+        echo "Missing CA certificate field in OVF template, which is required for Secure Boot support." >&2
+        exit 1
+      fi
+    fi
+  fi
+}
+
+get_data_image() {
+  local fmt input_dir filename_prefix
+  fmt="${1:?}"
+  input_dir="${2:?}"
+  filename_prefix="${3:?}"
+
+  if [[ "${fmt}" == "raw" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.img.lz4"
+  elif [[ "${fmt}" == "qcow2" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.qcow2"
+  elif [[ "${fmt}" == "vmdk" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.vmdk"
+  fi
+}
```

Comment:
Without seeing the caller for this function, it's hard to review but it does seem a bit strange. The caller has almost all of the information (because `${filename_prefix}${DATA_IMAGE_SUFFIX}` is the same as the `data_image_name` we set for them.

Reply:
I could save this for the `img2img` PR, but weirdness comes from our choice of formats. "raw" == "img.lz4", so we can't just do `${INPUT_DIR}/${FILENAME_PREFIX}${DATA_IMAGE_SUFFIX}.${OUTPUT_FMT}`. I could do `${INPUT_DIR}/${FILENAME_PREFIX}${DATA_IMAGE_SUFFIX}.${OUTPUT_FMT/raw/img.lz4}`, but ideally filename prefix and data image suffix will be in imghelper and not the caller. This seemed the most readable when factoring in the above.